### ISSUE TITLE
irrefutable_let_patterns should ignore `#[non_exhaustive]` structs matching

### ISSUE THREAD

**Issue by mathstuf** [rust-lang/rust#72469]
I tried this code:

```rust
#[non_exhaustive]
enum Single {
    Variant,
}

fn irrefutable(single: Single) {
    if let Single::Variant = single {
        // Some code.
    } else {
        panic!();
    }
}
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9efe847b2cf376a699323c2e2399cedd

I expected to see this happen:

No warning about the binding.

Instead, this happened:

An `irrefutable_let_patterns` warning is raised.

### Meta

Playground stable (1.43.1)

**Comment by jonas-schievink** [comment]
I think this is intentional, as `#[non_exhaustive]` only affects downstream crates

**Comment by mathstuf** [comment]
Hmm. What's the best way to ensure I get this error in a test then while being robust against future variants being added? I guess the warning could be suppressed for the affected tests. Should tests be considered downstream maybe? Though that sounds like a much bigger change.

**Comment by Nemo157** [comment]
External `tests/foobar.rs` feature tests and doctests are downstream and will need to obey `non_exhaustive`, only internal unit tests are part of the crate and will ignore it.

**Comment by jonas-schievink** [comment]
Closing as expected behavior.
